node/src/components/chain_synchronizer/operations.rs

```diff
@@ -2240,21 +2247,37 @@ async fn fetch_and_store_deploys<REv>(
     ctx: &ChainSyncContext<'_, REv>,
 ) -> Result<Vec<Deploy>, Error>
 where
-    REv: From<StorageRequest> + From<FetcherRequest<Deploy>> + From<NetworkInfoRequest>,
+    REv: From<StorageRequest>
+        + From<FetcherRequest<Deploy>>
+        + From<NetworkInfoRequest>
+        + From<ControlAnnouncement>,
 {
     let start_instant = Timestamp::now();
 
-    let hashes: Vec<_> = hashes.cloned().collect();
-    let mut deploys: Vec<Deploy> = Vec::with_capacity(hashes.len());
-    let mut stream = futures::stream::iter(hashes)
-        .map(|hash| fetch_and_store_deploy(hash, ctx))
+    let mut ordered_deploys = BTreeMap::new();
+
+    let mut stream = futures::stream::iter(hashes.cloned().enumerate())
+        .map(|(index, hash)| async move { (index, fetch_and_store_deploy(hash, ctx).await) })
         .buffer_unordered(ctx.config.max_parallel_deploy_fetches());
-    while let Some(result) = stream.next().await {
+    while let Some((index, result)) = stream.next().await {
         let deploy = result?;
         trace!("fetched {:?}", deploy);
-        deploys.push(*deploy);
+        if let Some(deploy) = ordered_deploys.insert(index, *deploy) {
```

Comment:
I am not a fan of this. It's a lot of code that is essentially checking that `enumerate` works correctly :/

Reply:
Can we not replace all of this code, starting from `stream`, with something like this? (I'm assuming `hashes` contains what we need in order):

```rust
let jobs: FuturesOrdered<_> = hashes.into_iter().map(|hash| fetch_and_store_deploy(hash, ctx));
let ordered_deploys: Vec<_> = jobs.collect().await;
```

Reply:
Ordering thousands of deploys? I would be worried about performance of `PartialEq, Eq` for `Deploy`.

Reply:
Hmm, I would not be concerned about that at all:

1. Ordering a few thousand items should be quick, especially if our ordering is by hash.
2. We actually don't even order anything there :) The `Ordered` part is just keeping the already existing order.

Reply:
There may be merit to using `buffered` instead of `buffer_unordered` instead as well.

Reply:
```rust
let mut stream = futures::stream::iter(hashes.cloned())
  .map(|hash| async move { fetch_and_store_deploy(hash, ctx).await })
  .buffered(ctx.config.max_parallel_deploy_fetches());
        
let deploys = stream.collect().await;
ctx.metrics
    .observe_fetch_deploys_duration_seconds(start_instant);
Ok(deploys)
```